doc/input-validation.md

```diff
@@ -0,0 +1,50 @@
+# Input Data Validation
+
+When we receive input data from an upstream API, we perform input validation to:
+
+- Ensure we won't throw a runtime error trying to render a badge
+- Ensure we won't render badges with spurious or unexpected output e.g: ![](https://img.shields.io/badge/version-null-blue) ![](https://img.shields.io/badge/coverage-NaN%25-red) ![](https://img.shields.io/badge/build-undefined-red) ![](https://img.shields.io/badge/coverage---10%25-critical) etc
+- Express and document our understanding of the input data
+
+## Writing schemas and validation
+
+- The default validation mechanism should be to use [Joi](https://github.com/sideway/joi) to define a schema for the input data. Validation against Joi schemas is implemented in the base classes every integration inherits from. Sometimes additional manual validation is needed which can't be covered by Joi and plugins in which case we implement it by hand.
```

Comment:
The reason why I picked the word "integration" here is because when I'm being careful, I think the correct words are:

* [NPM](https://www.npmjs.com/) is a **service**
* `class NpmVersion` is an **integration**
* ![](https://img.shields.io/npm/v/npm.svg) is a **badge**

but also I suspect we're probably not great about using that terminology throughout the docs. I've made a wording change based on your suggestion.

Reply:
> I think the correct words are:

    NPM is a service
    class NpmVersion is an integration
    is a badge

I know somewhere we've an open discussion about terminology so I'll try to find that and post my thoughts over there. Updated text LGTM